schoolRoster/schoolRoster.js

```diff
@@ -0,0 +1,123 @@
+
+        if (permission == 'y') {
+        promptName = require('prompt-sync')();
+        studentName = promptName('Enter student name:');
+        promptGrade = require('prompt-sync')();
+        studentGrade = Number(promptGrade('Enter student grade:'));
+
+        singleStudentInfo = {studentName: studentName, studentGrade: studentGrade};
+        //pushing objects as elements in array
+        schoolStudentInfo.push(singleStudentInfo);
+        
+        }else if(permission == 'n') {
+        console.log('Ok!');
+
+        } else {
+        console.log('Your input is wrong, enter y/n');
+        addStudentToGrade(); 
+        }
+    }
+ 
+    return studentName, studentGrade;
+}
+
+
+function getStudentsForGrade() {
+    promptPermission = require('prompt-sync')();
+    permission = promptPermission('Would you like to search for students in a grade?(y/n): ');
+    
+    if (permission == 'y') {
+        promptGrade = require('prompt-sync')();
+        studentGrade = Number(promptGrade('Enter grade to retrieve grade list:'));
+        console.log(`Student list for grade ${studentGrade} :`);
+        // loop through objects in array, find object with property = variable input
+        for(let i = 0; i<schoolStudentInfo.length; i++) {
+            if(schoolStudentInfo[i].studentGrade == studentGrade){
+                console.log(schoolStudentInfo[i].studentName);
+            }
+        }
+    } else if(permission == 'n') {
+        console.log('Ok!');
+    } else {
+        console.log('Your input is wrong, enter y/n');
+        getStudentsForGrade(); 
+    }
+
+    return studentGrade;
+}
+
+function getAllStudents() {
+    promptPermission = require('prompt-sync')();
+    permission = promptPermission('Would you like to display all grade lists?(y/n): ')
+    
+    if (permission == 'y') {
+        console.log(`Student list for all grades :`);
+        let baseNum = 1;
+        // limit is length of schoolStudentInfo because at the very least, only 1 student per grade
+        for (let i = 0; i<schoolStudentInfo.length; i++) {
+            // increment of baseNum only up to 12 i.e grade 12
+            if(baseNum < 13) {
+                console.log(`GRADE ${baseNum}: `);
+                //find student grade == baseNum
+                if (schoolStudentInfo[i].studentGrade === baseNum){
+                    console.log(schoolStudentInfo[i].studentName);
+                }
+                //loop through other  objects = baseNum ,and display all besides student in above for loop
+                for(let t = 0; t<schoolStudentInfo.length; t++){
+                    if(schoolStudentInfo[t].studentGrade == baseNum && schoolStudentInfo[i].studentName !== schoolStudentInfo[t].studentName){
+                    console.log(schoolStudentInfo[t].studentName);
+                    } 
+                }
+                //update baseNum for next loop and loop w/n loop
+                baseNum += 1;
+            }
+    
+        }
```

Comment:
Another implementation using the [.filter()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/filter), [.map() ](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map)and .[forEach()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach) functions could be this:

```suggestion
        //looping over all possible grades
        for (let grade = 1; grade<13; grade++) {
                console.log(`GRADE ${grade}: `);
                //find students in grade == grade
                // filter only students in specific grade. Returns a list of students
                const studentsInGrade = schoolStudentInfo.filter((student) => student.studentGrade === grade)
                //create a list of student names
                const studentNamesInGrade = studentsInGrade.map((student) => student.studentName)
                studentNamesInGrade.forEach((studentName) => {
                     console.log(studentName)
                 }
        }
```

Reply:
I'll definitely need to go through these functions to get a better understanding.